code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm

```diff
@@ -141,100 +143,90 @@
+		update_appearance()
 
-	var/heat_amount = temperature_delta * (main_heat_capacity * heat_capacity / (main_heat_capacity + heat_capacity))
-	var/efficiency = 1
-	var/temperature_difference = 0
-	var/skip_tick = TRUE
-	if(!use_enviroment_heat && main_port.total_moles() > 0.01)
-		if(cooling && thermal_exchange_port.total_moles() > 0.01 && nodes[2] && (thermal_exchange_port.temperature <= THERMOMACHINE_SAFE_TEMPERATURE || !safeties))
-			thermal_exchange_port.temperature = max(thermal_exchange_port.temperature + heat_amount / thermal_heat_capacity + motor_heat / thermal_heat_capacity, TCMB)
-		else if(cooling && (!thermal_exchange_port.total_moles() || !nodes[2]))
-			skipping_work = skip_tick
-			update_appearance()
-			update_parents()
-			return
-		if(thermal_exchange_port.temperature > THERMOMACHINE_SAFE_TEMPERATURE && safeties)
-			on = FALSE
-			visible_message("<span class='warning'>The thermal exchange port's temperature has reached critical levels, shutting down...</span>")
-			update_appearance()
-			return
-		else if(thermal_exchange_port.temperature > THERMOMACHINE_SAFE_TEMPERATURE && !safeties)
-			if((REALTIMEOFDAY - lastwarning) / 5 >= WARNING_DELAY)
-				lastwarning = REALTIMEOFDAY
-				visible_message("<span class='warning'>The thermal exchange port's temperature has reached critical levels!</span>")
-				if(check_explosion(thermal_exchange_port.temperature))
-					explode()
-					return PROCESS_KILL //we dying anyway, so let's stop processing
-		temperature_difference = thermal_exchange_port.temperature - main_port.temperature
-		temperature_difference = cooling ? temperature_difference : 0
-		if(temperature_difference > 0)
-			efficiency = max(1 - log(10, temperature_difference) * 0.08, 0.65)
-		main_port.temperature = max(main_port.temperature - (heat_amount * efficiency)/ main_heat_capacity + motor_heat / main_heat_capacity, TCMB)
-		skip_tick = FALSE
-	if(use_enviroment_heat && main_port.total_moles() > 0.01)
-		var/enviroment_efficiency = 1
-		if(cooling && enviroment.total_moles() > 0.01 && (thermal_exchange_port.temperature <= THERMOMACHINE_SAFE_TEMPERATURE || !safeties))
-			var/enviroment_heat_capacity = enviroment.heat_capacity()
-			if(enviroment.total_moles())
-				enviroment_efficiency = clamp(log(1.55, enviroment.total_moles()) * 0.15, 0.65, 1)
-			enviroment.temperature = max(enviroment.temperature + heat_amount / enviroment_heat_capacity, TCMB)
-			air_update_turf(FALSE, FALSE)
-		else if(cooling && !enviroment.total_moles())
-			skipping_work = skip_tick
-			update_appearance()
-			update_parents()
+	skipping_work = FALSE
+
+	if (main_port.total_moles() < 0.01)
+		skipping_work = TRUE
+		return
+
+	// Efficiency should be a proc level variable, but we need it for the ui. 
+	// This is to reset the value when we are heating.
+	efficiency = 1
+
+	if(cooling)
+		var/datum/gas_mixture/exchange_target
+		// Exchange target is the thing we are paired with, be it enviroment or the red port. 
```

Comment:
Why is this all locked behind cooling? Don't we want to use this logic all of the time? or am I missing something here

Reply:
The thermomachine isn't meant to play with the second port on heating mode. That giant block of code is there just to count the efficiency and heat up the exchange part when cooling.